serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,282 @@
+				// Failing to close response bodies can present as a memory leak so it's
+				// important to surface it.
+				c.logger.Warn(fmt.Sprintf("unable to close response body: %s", err))
+			}
+		}
+	}()
+
+	// Check for success.
+	switch resp.StatusCode {
+	case 200, 201, 202:
+		// Pass.
+	case 401, 403:
+		// Perhaps the token from our bearer token file has been refreshed.
+		config, err := inClusterConfig()
+		if err != nil {
+			return false, err
+		}
+		if config.BearerToken == c.config.BearerToken {
+			// It's the same token.
+			return false, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+		}
+		c.config = config
+		// Continue to try again, but return the error too in case the caller would rather read it out.
+		return true, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	case 404:
+		return false, ErrNotFound
+	case 500, 502, 503, 504:
+		// Could be transient.
+		return true, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	default:
+		// Unexpected.
+		return false, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	}
+
+	// We only arrive here with success.
+	// If we're not supposed to read out the body, we have nothing further
+	// to do here.
+	if ptrToReturnObj == nil {
+		return false, nil
+	}
+
+	// Attempt to read out the body into the given return object.
+	return false, json.NewDecoder(resp.Body).Decode(ptrToReturnObj)
+}
+
+type Pod struct {
+	Metadata *Metadata `json:"metadata,omitempty"`
+}
+
+type Metadata struct {
+	Name string `json:"name,omitempty"`
+
+	// This map will be nil if no "labels" key was provided.
+	// It will be populated but have a length of zero if the
+	// key was provided, but no values.
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+type PatchOperation int
+
+const (
+	// When adding support for more PatchOperations in the future,
+	// DO NOT alphebetize them because it will change the underlying
+	// int representing a user's intent. If that's stored anywhere,
+	// it will cause storage reads to map to the incorrect operation.
+	Unset PatchOperation = iota
+	Add
+	Replace
+)
+
+func Parse(s string) PatchOperation {
+	switch s {
+	case "add":
+		return Add
+	case "replace":
+		return Replace
+	default:
+		return Unset
+	}
+}
+
+func (p PatchOperation) String() string {
+	switch p {
+	case Unset:
+		// This is an invalid choice, and will be shown on a patch
+		// where the PatchOperation is unset. That's because ints
+		// default to 0, and Unset corresponds to 0.
+		return "unset"
+	case Add:
+		return "add"
+	case Replace:
+		return "replace"
+	default:
+		// Should never arrive here.
+		return ""
+	}
+}
+
+type Patch struct {
+	Operation PatchOperation
+	Path      string
+	Value     interface{}
+}
+
+// sanitizedDebuggingInfo converts an http response to a string without
+// including its headers to avoid leaking authorization
+// headers.
```

Comment:
Nit: This could probably be moved up to the end of the last line since the first comment line is longer.

Reply:
Fixed in https://github.com/hashicorp/vault/pull/8249/commits/c211b87183fb94eec61b37f8c486202ef355ce52